File: Coding Problems & My Solutions/Maximum Product Subarray/Maximum_Product_Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int hi = nums[0];
        int lo = nums[0];
        int max = hi;
        for (int i = 1; i < nums.length; i++){
            int cur = nums[i];
            int oldhi = hi;
            hi = Math.max(cur, Math.max(cur*hi, cur*lo));
            lo = Math.min(cur, Math.min(cur*oldhi, cur*lo));
            max = Math.max(max, hi);
        }
        return max;
    }
}

File: Coding Problems & My Solutions/Maximum Product Subarray/Maximum_Product_Subarray.swift
class Solution {
    func maxProduct(_ nums: [Int]) -> Int {
        var ans = Int.min
        var curMax = 1
        var curMin = 1
        for num in nums {
            var tempMax = curMax
            curMax = max(num * curMax, num * curMin, num)
            curMin = min(num * tempMax, num * curMin, num)
            ans = max(ans, curMax)
        }
        return ans
    }
}

class Solution {
    func maxProduct(_ nums: [Int]) -> Int {
        var ans = Int.min
        var currentMax = 1
        var currentMin = 1
        for num in nums {
            var tempMax = currentMax
            currentMax = max(num, num * currentMax, num * currentMin)
            currentMin = min(num, num * tempMax, num * currentMin)
            ans = max(ans, num, currentMax)
        }
        return ans
    }
}